examples/keygen/keygen.c

```diff
@@ -0,0 +1,152 @@
+#include <wolftpm/tpm2_wrap.h>
+
+#include <examples/keygen/keygen.h>
+#include <examples/tpm_io.h>
+#include <examples/tpm_test.h>
+
+#include <stdio.h>
+
+
+/******************************************************************************/
+/* --- BEGIN TPM Keygen Example -- */
+/******************************************************************************/
+
+int TPM2_Keygen_Example(void* userCtx, int argc, char *argv[])
+{
+    int rc;
+    WOLFTPM2_DEV dev;
+    WOLFTPM2_KEY storage; /* SRK */
+    WOLFTPM2_KEYBLOB newKey;
+    TPMT_PUBLIC publicTemplate;
+    TPMS_AUTH_COMMAND session[MAX_SESSION_NUM];
+    TPMI_ALG_PUBLIC alg = TPM_ALG_RSA; /* TPM_ALG_ECC */
+
+#if !defined(WOLFTPM2_NO_WOLFCRYPT) && !defined(NO_FILESYSTEM)
+    XFILE f;
+    const char* outputFile = "keyblob.bin";
+
+    if (argc >= 2) {
+        outputFile = argv[1];
+    }
+#endif
+    XMEMSET(session, 0, sizeof(session));
+    XMEMSET(&storage, 0, sizeof(storage));
+    XMEMSET(&newKey, 0, sizeof(newKey));
+
+    printf("TPM2.0 Key generation example\n");
+    rc = wolfTPM2_Init(&dev, TPM2_IoCb, userCtx);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("\nwolfTPM2_Init failed\n");
+        goto exit;
+    }
+    printf("wolfTPM2_Init: success\n\n");
+
+
+    /* Define the default session auth that has NULL password */
+    session[0].sessionHandle = TPM_RS_PW;
+    session[0].auth.size = 0;
+    TPM2_SetSessionAuth(session);
+
+    /* See if SRK already exists */
+    rc = wolfTPM2_ReadPublicKey(&dev, &storage, TPM2_DEMO_STORAGE_KEY_HANDLE);
+    if (rc != 0) {
+        printf("Loading SRK: Storage failed 0x%x: %s\n", rc,
+            TPM2_GetRCString(rc));
+        goto exit;
+    }
+    printf("Loading SRK: Storage 0x%x (%d bytes)\n",
+        (word32)storage.handle.hndl, storage.pub.size);
+
+    /* set session for authorization of the storage key */
+    session[0].auth.size = sizeof(gStorageKeyAuth)-1;
+    XMEMCPY(session[0].auth.buffer, gStorageKeyAuth, session[0].auth.size);
+
+    /* Create new key */
+    if (alg == TPM_ALG_RSA) {
+        rc = wolfTPM2_GetKeyTemplate_RSA_AIK(&publicTemplate);
+    }
+    else if (alg == TPM_ALG_ECC) {
+        rc = wolfTPM2_GetKeyTemplate_ECC_AIK(&publicTemplate);
+    }
+    else {
+        rc = BAD_FUNC_ARG;
+        goto exit;
+    }
+    printf("Creating new %s key...\n", TPM2_GetAlgName(alg));
+    rc = wolfTPM2_CreateKey(&dev, &newKey, &storage.handle,
+        &publicTemplate, (const byte*)gAiKeyAuth, sizeof(gAiKeyAuth)-1);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("wolfTPM2_CreateKey failed\n");
+        goto exit;
+    }
+    printf("Created new key (pub %d, priv %d bytes)\n",
+        newKey.pub.size, newKey.priv.size);
+
+    /* Save key as encrypted blob to the disk */
+#if !defined(WOLFTPM2_NO_WOLFCRYPT) && !defined(NO_FILESYSTEM)
+    f = XFOPEN(outputFile, "wb");
+    if (f != XBADFILE) {
+        XFWRITE(&newKey.pub, 1, sizeof(newKey.pub), f);
+        XFWRITE(&newKey.priv, 1, sizeof(newKey.priv), f);
```

Comment:
This doesn't have to write the entire priv buffer, just `newKey.priv.size + sizeof(UINT16)`.